src/render/source.rs

```diff
@@ -0,0 +1,133 @@
+//! Source for for GLSL shaders used by the renderer.
+
+use data;
+use util;
+
+use std::{io, ops, str};
+use std::borrow::Borrow;
+use std::path::Path;
+
+/// Source code for a single GLSL shader.
+#[derive(Clone, Debug)]
+pub struct Source(pub(crate) String);
+
+impl ops::Deref for Source {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        self.0.as_bytes()
+    }
+}
+
+impl Source {
+    fn preprocess<P: AsRef<Path>>(
+        root: P,
+        code: &str,
+    ) -> io::Result<String> {
+        let root = root.as_ref();
+        let mut new_code = String::new();
+        for line in code.lines() {
+            if line.starts_with("#include") {
+                for dep_name in line.split(' ').skip(1) {
+                    match dep_name {
+                        "locals" | "lights" | "globals" => {
```

Comment:
hmm, I'd prefer us to have a solid distinction of the built-in versus custom includes in shaders
something like `locals` for our shaders and `/locals` for user shaders

Reply:
Is that not covered already? :confused: